File: src/Storage/FileStorage.php
<?php

declare(strict_types=1);

namespace MiniS3\Storage;

use FilesystemIterator;
use RecursiveDirectoryIterator;
use RecursiveIteratorIterator;
use RuntimeException;

final class FileStorage
{
    private const MULTIPART_ROOT = '.multipart';

    public function __construct(private readonly string $dataDir)
    {
    }

    public function ensureDataDirExists(): void
    {
        if (is_dir($this->dataDir)) {
            return;
        }

        if (!mkdir($this->dataDir, 0777, true) && !is_dir($this->dataDir)) {
            throw new RuntimeException('Failed to create data directory');
        }
    }

    public function listFiles(string $bucket, string $prefix = ''): array
    {
        $dir = $this->dataDir . '/' . $bucket;
        if (!is_dir($dir)) {
            return [];
        }

        $files = [];
        $iterator = new RecursiveIteratorIterator(new RecursiveDirectoryIterator($dir, FilesystemIterator::SKIP_DOTS));

        foreach ($iterator as $file) {
            if (!$file->isFile()) {
                continue;
            }

            if (str_starts_with($file->getFilename(), '.')) {
                continue;
            }

            $relativePath = substr($file->getPathname(), strlen($dir) + 1);
            if ($prefix !== '' && !str_starts_with($relativePath, $prefix)) {
                continue;
            }

            $files[] = [
                'key' => $relativePath,
                'size' => $file->getSize(),
                'timestamp' => $file->getMTime(),
            ];
        }

        return $files;
    }

    public function objectPath(string $bucket, string $key): string
    {
        return $this->dataDir . '/' . $bucket . '/' . $key;
    }

    public function objectExists(string $bucket, string $key): bool
    {
        return is_file($this->objectPath($bucket, $key));
    }

    public function putObjectFromInput(string $bucket, string $key): void
    {
        $filePath = $this->objectPath($bucket, $key);
        $this->ensureDirectory(dirname($filePath));
        $this->copyInputToAtomicFile($filePath);
    }

    public function deleteObject(string $bucket, string $key): void
    {
        $filePath = $this->objectPath($bucket, $key);
        if (is_file($filePath)) {
            unlink($filePath);
        }
    }

    public function createMultipartUpload(string $bucket, string $key, string $uploadId): void
    {
        $uploadDir = $this->multipartDir($bucket, $key, $uploadId);
        $this->ensureDirectory($uploadDir);
    }

    public function multipartDir(string $bucket, string $key, string $uploadId): string
    {
        return $this->dataDir
            . '/' . self::MULTIPART_ROOT
            . '/' . $bucket
            . '/' . $this->keyNamespace($key)
            . '/' . $uploadId;
    }

    public function multipartDirExists(string $bucket, string $key, string $uploadId): bool
    {
        return is_dir($this->multipartDir($bucket, $key, $uploadId));
    }

    public function putMultipartPartFromInput(string $bucket, string $key, string $uploadId, int $partNumber): string
    {
        $uploadDir = $this->multipartDir($bucket, $key, $uploadId);
        if (!is_dir($uploadDir)) {
            throw new RuntimeException('Upload ID not found');
        }

        $partPath = $uploadDir . '/' . $partNumber;
        $this->copyInputToAtomicFile($partPath);

        return $partPath;
    }

    public function completeMultipartUpload(string $bucket, string $key, string $uploadId, array $partNumbers): void
    {
        $uploadDir = $this->multipartDir($bucket, $key, $uploadId);
        if (!is_dir($uploadDir)) {
            throw new RuntimeException('Upload ID not found');
        }

        $filePath = $this->objectPath($bucket, $key);
        $this->ensureDirectory(dirname($filePath));

        $tmpPath = $this->createTempPath(dirname($filePath), '.obj-');
        $out = fopen($tmpPath, 'wb');
        if ($out === false) {
            throw new RuntimeException('Failed to open destination file');
        }

        try {
            foreach ($partNumbers as $partNumber) {
                $partPath = $uploadDir . '/' . $partNumber;
                if (!is_file($partPath)) {
                    throw new RuntimeException('Part file missing: ' . $partNumber);
                }

                $in = fopen($partPath, 'rb');
                if ($in === false) {
                    throw new RuntimeException('Failed to open multipart part: ' . $partNumber);
                }

                $copied = stream_copy_to_stream($in, $out);
                fclose($in);
                if ($copied === false) {
                    throw new RuntimeException('Failed to merge multipart part: ' . $partNumber);
                }
            }
        } catch (RuntimeException $e) {
            fclose($out);
            @unlink($tmpPath);
            throw $e;
        }

        fclose($out);

        if (!rename($tmpPath, $filePath)) {
            @unlink($tmpPath);
            throw new RuntimeException('Failed to finalize destination file');
        }
    }

    public function cleanupMultipartUpload(string $bucket, string $key, string $uploadId): void
    {
        $uploadDir = $this->multipartDir($bucket, $key, $uploadId);
        $this->deleteDirectory($uploadDir);

        $keyRoot = dirname($uploadDir);
        $bucketRoot = dirname($keyRoot);
        $multipartRoot = $this->dataDir . '/' . self::MULTIPART_ROOT;

        $this->removeIfEmptyDirectory($keyRoot);
        $this->removeIfEmptyDirectory($bucketRoot);
        $this->removeIfEmptyDirectory($multipartRoot);
    }

    public function abortMultipartUpload(string $bucket, string $key, string $uploadId): void
    {
        $this->cleanupMultipartUpload($bucket, $key, $uploadId);
    }

    private function copyInputToAtomicFile(string $targetPath): void
    {
        $inputStream = PHP_SAPI === 'cli' ? 'php://stdin' : 'php://input';
        $input = fopen($inputStream, 'rb');
        if ($input === false) {
            throw new RuntimeException('Failed to read request body');
        }

        $tmpPath = $this->createTempPath(dirname($targetPath), '.upload-');
        $output = fopen($tmpPath, 'wb');
        if ($output === false) {
            fclose($input);
            throw new RuntimeException('Failed to write file');
        }

        $copied = stream_copy_to_stream($input, $output);

        fclose($output);
        fclose($input);

        if ($copied === false) {
            @unlink($tmpPath);
            throw new RuntimeException('Failed to write file');
        }

        if (!rename($tmpPath, $targetPath)) {
            @unlink($tmpPath);
            throw new RuntimeException('Failed to finalize file');
        }
    }

    private function ensureDirectory(string $dir): void
    {
        if (is_dir($dir)) {
            return;
        }

        if (!mkdir($dir, 0777, true) && !is_dir($dir)) {
            throw new RuntimeException('Failed to create directory: ' . $dir);
        }
    }

    private function deleteDirectory(string $path): void
    {
        if (!file_exists($path)) {
            return;
        }

        if (is_file($path) || is_link($path)) {
            unlink($path);
            return;
        }

        $iterator = new FilesystemIterator($path, FilesystemIterator::SKIP_DOTS);
        foreach ($iterator as $item) {
            $this->deleteDirectory($item->getPathname());
        }

        rmdir($path);
    }

    private function isDirectoryEmpty(string $dir): bool
    {
        $iterator = new FilesystemIterator($dir, FilesystemIterator::SKIP_DOTS);

        return !$iterator->valid();
    }

    private function keyNamespace(string $key): string
    {
        if ($key === '') {
            return '_root';
        }

        return hash('sha256', $key);
    }

    private function createTempPath(string $directory, string $prefix): string
    {
        $tmpPath = tempnam($directory, $prefix);
        if ($tmpPath === false) {
            throw new RuntimeException('Failed to create temporary file');
        }

        return $tmpPath;
    }

    private function removeIfEmptyDirectory(string $dir): void
    {
        if (is_dir($dir) && $this->isDirectoryEmpty($dir)) {
            @rmdir($dir);
        }
    }
}

File: go/internal/storage/file_storage.go
package storage

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
)

const multipartRoot = ".multipart"

// FileStorage implements filesystem-backed object storage, mirroring
// MiniS3\Storage\FileStorage from the PHP reference.
type FileStorage struct {
	dataDir string
}

// New creates a FileStorage rooted at dataDir.
func New(dataDir string) *FileStorage {
	return &FileStorage{dataDir: dataDir}
}

// DataDir returns the configured data directory.
func (s *FileStorage) DataDir() string {
	return s.dataDir
}

// EnsureDataDirExists creates the data directory if it doesn't exist.
func (s *FileStorage) EnsureDataDirExists() error {
	return os.MkdirAll(s.dataDir, 0777)
}

// FileInfo holds metadata about a stored object.
type FileInfo struct {
	Key       string
	Size      int64
	Timestamp int64
}

// ListFiles recursively lists files in a bucket, optionally filtered by prefix.
// Hidden files (starting with ".") are excluded, matching PHP behavior.
func (s *FileStorage) ListFiles(bucket, prefix string) ([]FileInfo, error) {
	dir := filepath.Join(s.dataDir, bucket)
	entries, err := os.ReadDir(dir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}

	var files []FileInfo
	err = filepath.WalkDir(dir, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			return nil
		}
		name := d.Name()
		if strings.HasPrefix(name, ".") {
			return nil
		}

		rel, err := filepath.Rel(dir, path)
		if err != nil {
			return nil
		}
		// Use forward slashes for S3 key compatibility.
		rel = filepath.ToSlash(rel)

		if prefix != "" && !strings.HasPrefix(rel, prefix) {
			return nil
		}

		info, err := d.Info()
		if err != nil {
			return nil
		}
		files = append(files, FileInfo{
			Key:       rel,
			Size:      info.Size(),
			Timestamp: info.ModTime().Unix(),
		})
		return nil
	})
	// We already read the dir; if WalkDir fails because the dir doesn't
	// exist, that's fine (already handled above). But WalkDir may re-stat.
	if err != nil && !os.IsNotExist(err) {
		return nil, err
	}
	_ = entries
	return files, nil
}

// ObjectPath returns the full filesystem path for a bucket/key pair.
func (s *FileStorage) ObjectPath(bucket, key string) string {
	return filepath.Join(s.dataDir, bucket, key)
}

// ObjectExists reports whether the object exists as a regular file.
func (s *FileStorage) ObjectExists(bucket, key string) bool {
	info, err := os.Stat(s.ObjectPath(bucket, key))
	return err == nil && !info.IsDir()
}

// ObjectMetadata returns size and MIME type for an object, or nil if not found.
type ObjectMetadata struct {
	Size     int64
	MimeType string
}

func (s *FileStorage) ObjectMetadata(bucket, key string) (*ObjectMetadata, error) {
	path := s.ObjectPath(bucket, key)
	info, err := os.Stat(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	return &ObjectMetadata{
		Size:     info.Size(),
		MimeType: s.detectMimeType(path),
	}, nil
}

// OpenObjectReadStream opens the object for reading.
func (s *FileStorage) OpenObjectReadStream(bucket, key string) (*os.File, error) {
	return os.Open(s.ObjectPath(bucket, key))
}

// PutObject writes the object from the provided reader (atomic: temp + rename).
func (s *FileStorage) PutObject(bucket, key string, body io.Reader) error {
	path := s.ObjectPath(bucket, key)
	if err := s.ensureDir(filepath.Dir(path)); err != nil {
		return err
	}
	return s.atomicWrite(path, body)
}

// DeleteObject removes the object file if it exists.
func (s *FileStorage) DeleteObject(bucket, key string) error {
	path := s.ObjectPath(bucket, key)
	err := os.Remove(path)
	if err != nil && !os.IsNotExist(err) {
		return err
	}
	return nil
}

// CreateMultipartUpload creates the multipart staging directory.
func (s *FileStorage) CreateMultipartUpload(bucket, key, uploadID string) error {
	uploadDir := s.MultipartDir(bucket, key, uploadID)
	return s.ensureDir(uploadDir)
}

// MultipartDir returns the directory path for a multipart upload.
func (s *FileStorage) MultipartDir(bucket, key, uploadID string) string {
	return filepath.Join(s.dataDir, multipartRoot, bucket, s.keyNamespace(key), uploadID)
}

// MultipartDirExists reports whether the multipart upload directory exists.
func (s *FileStorage) MultipartDirExists(bucket, key, uploadID string) bool {
	info, err := os.Stat(s.MultipartDir(bucket, key, uploadID))
	return err == nil && info.IsDir()
}

// PutMultipartPart writes a multipart part from the reader and returns its path.
func (s *FileStorage) PutMultipartPart(bucket, key, uploadID string, partNumber int, body io.Reader) (string, error) {
	uploadDir := s.MultipartDir(bucket, key, uploadID)
	if _, err := os.Stat(uploadDir); err != nil {
		return "", fmt.Errorf("upload ID not found")
	}
	partPath := filepath.Join(uploadDir, fmt.Sprintf("%d", partNumber))
	if err := s.atomicWrite(partPath, body); err != nil {
		return "", err
	}
	return partPath, nil
}

// CompleteMultipartUpload assembles parts (in order) into the final object.
func (s *FileStorage) CompleteMultipartUpload(bucket, key, uploadID string, partNumbers []int) error {
	uploadDir := s.MultipartDir(bucket, key, uploadID)
	if _, err := os.Stat(uploadDir); err != nil {
		return fmt.Errorf("upload ID not found")
	}

	filePath := s.ObjectPath(bucket, key)
	if err := s.ensureDir(filepath.Dir(filePath)); err != nil {
		return err
	}

	tmpPath := s.createTempPath(filepath.Dir(filePath), ".obj-")
	out, err := os.Create(tmpPath)
	if err != nil {
		return fmt.Errorf("failed to open destination file: %w", err)
	}

	for _, partNumber := range partNumbers {
		partPath := filepath.Join(uploadDir, fmt.Sprintf("%d", partNumber))
		in, err := os.Open(partPath)
		if err != nil {
			out.Close()
			os.Remove(tmpPath)
			return fmt.Errorf("part file missing: %d", partNumber)
		}
		_, err = io.Copy(out, in)
		in.Close()
		if err != nil {
			out.Close()
			os.Remove(tmpPath)
			return fmt.Errorf("failed to merge multipart part: %d", partNumber)
		}
	}
	out.Close()

	if err := os.Rename(tmpPath, filePath); err != nil {
		os.Remove(tmpPath)
		return fmt.Errorf("failed to finalize destination file: %w", err)
	}
	return nil
}

// CleanupMultipartUpload removes the multipart upload directory and cleans
// up empty parent directories.
func (s *FileStorage) CleanupMultipartUpload(bucket, key, uploadID string) error {
	uploadDir := s.MultipartDir(bucket, key, uploadID)
	if err := s.deleteDirectory(uploadDir); err != nil && !os.IsNotExist(err) {
		return err
	}

	keyRoot := filepath.Dir(uploadDir)
	bucketRoot := filepath.Dir(keyRoot)
	multipartRootPath := filepath.Join(s.dataDir, multipartRoot)

	s.removeIfEmpty(keyRoot)
	s.removeIfEmpty(bucketRoot)
	s.removeIfEmpty(multipartRootPath)
	return nil
}

// AbortMultipartUpload is an alias for CleanupMultipartUpload.
func (s *FileStorage) AbortMultipartUpload(bucket, key, uploadID string) error {
	return s.CleanupMultipartUpload(bucket, key, uploadID)
}

func (s *FileStorage) atomicWrite(targetPath string, body io.Reader) error {
	tmpPath := s.createTempPath(filepath.Dir(targetPath), ".upload-")
	out, err := os.Create(tmpPath)
	if err != nil {
		return fmt.Errorf("failed to write file: %w", err)
	}

	_, err = io.Copy(out, body)
	closeErr := out.Close()
	if err != nil {
		os.Remove(tmpPath)
		return fmt.Errorf("failed to write file: %w", err)
	}
	if closeErr != nil {
		os.Remove(tmpPath)
		return fmt.Errorf("failed to write file: %w", closeErr)
	}

	if err := os.Rename(tmpPath, targetPath); err != nil {
		os.Remove(tmpPath)
		return fmt.Errorf("failed to finalize file: %w", err)
	}
	return nil
}

func (s *FileStorage) ensureDir(dir string) error {
	return os.MkdirAll(dir, 0777)
}

func (s *FileStorage) deleteDirectory(path string) error {
	info, err := os.Stat(path)
	if err != nil {
		return err
	}
	if !info.IsDir() {
		return os.Remove(path)
	}
	entries, err := os.ReadDir(path)
	if err != nil {
		return err
	}
	for _, entry := range entries {
		if err := s.deleteDirectory(filepath.Join(path, entry.Name())); err != nil {
			return err
		}
	}
	return os.Remove(path)
}

func (s *FileStorage) keyNamespace(key string) string {
	if key == "" {
		return "_root"
	}
	h := sha256.Sum256([]byte(key))
	return hex.EncodeToString(h[:])
}

func (s *FileStorage) createTempPath(directory, prefix string) string {
	f, err := os.CreateTemp(directory, prefix)
	if err != nil {
		// Fall back to a random name if CreateTemp fails.
		return filepath.Join(directory, prefix+randomSuffix())
	}
	name := f.Name()
	f.Close()
	return name
}

func (s *FileStorage) removeIfEmpty(dir string) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return
	}
	if len(entries) == 0 {
		os.Remove(dir)
	}
}

func (s *FileStorage) detectMimeType(path string) string {
	// Open the file and sniff the first 512 bytes, similar to PHP's
	// mime_content_type but using Go's mime detection.
	f, err := os.Open(path)
	if err != nil {
		return "application/octet-stream"
	}
	defer f.Close()

	buf := make([]byte, 512)
	n, err := f.Read(buf)
	if err != nil && err != io.EOF {
		return "application/octet-stream"
	}

	mimeType := http.DetectContentType(buf[:n])
	if mimeType == "" {
		return "application/octet-stream"
	}
	// http.DetectContentType may return a charset suffix (e.g.
	// "text/plain; charset=utf-8"). Strip it to match PHP's
	// mime_content_type which returns only the MIME type.
	if i := strings.Index(mimeType, ";"); i != -1 {
		mimeType = strings.TrimSpace(mimeType[:i])
	}
	return mimeType
}

func randomSuffix() string {
	return fmt.Sprintf("%d", os.Getpid())
}
